polynote-kernel/src/main/scala/polynote/kernel/Result.scala

```diff
@@ -190,16 +212,26 @@ object Result {
   implicit val codec: Codec[Result] = cachedImplicit
 
   def toCellUpdate(result: Result): NotebookCell => NotebookCell = {
-    // process carriage returns in the string
-    def collapseCrs(str: String): String = str.replaceAll("\\r\\n", "\n").replaceAll("[^\\n]*\\r", "")
+    // process carriage returns in the line – a carriage return deletes anything before it in the line, unless
+    // it's the last character before the linefeed (which must be the final character)
+    def collapseCrs(line: String): String = line.lastIndexOf('\r', line.length - 3) match {
+      case -1 => line
+      case n  => line.drop(n + 1)
+    }
 
     result match {
       case ClearResults() => _.copy(results = ShortList(Nil))
       case execInfo@ExecutionInfo(_, _) => cell => cell.copy(results = ShortList(cell.results :+ execInfo), metadata = cell.metadata.copy(executionInfo = Some(execInfo)))
-      case Output(contentType, str) =>
+      case Output("text/plain; rel=stdout", lines) =>
```

Comment:
TODO: will make `Output` abstract and have a dedicated leaf for stdout, text, and binary.